src/as2fm/scxml_converter/scxml_entries/type_utils.py

```diff
@@ -0,0 +1,180 @@
+
+
+class ScxmlStructDeclarationsContainer:
+    """
+    Object containing all the type-related information needed when generating the plain-SCXML model.
+
+    In this class, we hold all the information about:
+    * Data variables and related type
+    * Events and related types
+    * ROS Services and Actions
+    * Timer callbacks
+    """
+
+    def __init__(
+        self,
+        automaton_name: str,
+        data_model,
+        struct_definitions: Dict[str, XmlStructDefinition],
+    ):
+        self._automaton_name = automaton_name
+        self._data_model = data_model
+        self._struct_definitions = struct_definitions
+        self._type_per_variable: Dict[
+            str, Tuple[Union[XmlStructDefinition, str], Optional[ArrayInfo]]
+        ] = {}
+        for data_entry in self._data_model.get_data_entries():
+            variable_name: str = data_entry.get_name()
+            assert (
+                variable_name not in self._type_per_variable.keys()
+            ), f"Variable name {variable_name} must be unique."
+            data_type_def = data_entry.get_type_str()
+            if is_type_string_array(data_type_def):
+                data_type_single = get_type_string_of_array(data_type_def)
+                array_info: Optional[ArrayInfo] = get_array_info(data_type_def, False)
+            else:
+                data_type_single = data_type_def
+                array_info = None
+            if is_type_string_base_type(data_type_single):
+                self._type_per_variable[variable_name] = (data_type_def, array_info)
+            else:
+                data_type_struct = struct_definitions[data_type_single]
+                self._type_per_variable[variable_name] = (
+                    data_type_struct,
+                    array_info,
+                )
+
+    def get_data_type(
+        self, variable_name: str, elem
+    ) -> Tuple[Union[XmlStructDefinition, str], Optional[ArrayInfo]]:
+        """
+        Retrieve the type of a variable.
+
+        Used, when trying to assign to it.
+
+        e.g. `x` => ('int32', None)` <-- retrieve from dict
+        e.g. `xs` => ('int32', ArrayInfo)` <-- retrieve from dict
+        e.g. `xs[1]` => ('int32', None)` <- handle *specially*
+
+        e.g. `a_point` => (XmlStructDefinition(Point2D), None)` <-- retrieve from dict
+        e.g. `polygon.points` => `(XmlStructDefinition(Point2D), ArrayInfo)` <- ???
+        e.g. `polygon.points[1]` => `(XmlStructDefinition(Point2D), None)` <- ???
+        e.g. `polygon.points[1].x` => `(int32, None)` <- ???
+
+        """
+        access_trace = split_by_access(variable_name, elem)
+        return self._get_data_type_for_variable(access_trace, elem)
+
+    def _check_array_length_access(
+        self, access_trace: List[Union[str, Type[ArrayAccess]]], array_info: Optional[ArrayInfo]
+    ) -> bool:
+        """
+        Check if this expression relates to an array length access.
+        """
+        if array_info is not None:
```

Comment:
Add an assertion here, to ensure that those two conditions are holding when the keyword in `ARRAY_LENGTH_SUFFIX` is used.